tools: support cases where bootloader is in chunks

naveenkaje opened the pull request:
Bootloader can be placed beyond the application. Support this requirement.
With FEATURE_BOOTLOADER support, the bootloader can be placed at a
high address. Add support to the tools so that Application can be placed in
the available space before the beginning of the Bootloader.

### Pull request type
    [ ] Fix
    [ ] Refactor
    [ ] Target update
    [x] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

naveenkaje commented:
Tested with cloud client.

Log from build where BOOTLOADER for NRF52 is placed at 0x64000
```
0 2408
409600 478028
New offset for placing application 4096 last_address 409600 size 405504
rom_end 524288 rom_end - start 520192
Using ROM regions bootloader, header, application in this build.
  Region bootloader: size 0x75000, offset 0x0
  Region header: size 0x70, offset 0x75000
  Region application: size 0x63000, offset 0x1000
```
[output_bl_64000.txt](https://github.com/ARMmbed/mbed-os/files/2742909/output_bl_64000.txt)

Log with configuration where the bootloader, header and application are linearly placed
```
New offset for placing application 73856 last_address 524288 size 450432
rom_end 524288 rom_end - start 450432
Using ROM regions bootloader, header, application in this build.
  Region bootloader: size 0x12000, offset 0x0
  Region header: size 0x70, offset 0x12000
  Region application: size 0x6df80, offset 0x12080
```
[LinearPlacing.txt](https://github.com/ARMmbed/mbed-os/files/2742910/LinearPlacing.txt)

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -769,9 +769,9 @@ def _make_header_region(self, start, header_format, offset=None):
 
     @staticmethod
-    def _assign_new_offset(rom_start, start, new_offset, region_name):
+    def _assign_new_offset(rom_start, start, new_offset, region_name, override=False):
```
So:
 * Adding `override=True` skips the check in this function
 * This function does `rom_start + new_offset` and an overflow check

I think adding override to this function misses the point: this function exists for combining the overflow check with the new offset calculation.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -836,10 +836,23 @@ def _generate_bootloader_build(self, rom_memories):
                 start = self._assign_new_offset(
-                    rom_start, start, self.target.app_offset, "application")
-            yield Region("application", start, rom_end - start,
+                    rom_start, start, self.target.app_offset, "application", True)
```
This change will not protect users from the following scenario, which it was originally designed for:
Say you have the following:
 * A 16 K BL, end address is `0x4000`
 * Application configures it's offset to `0x3000`

This situation should result in an error. With this change, it might raise an exception from IntelHex which will not help the user. I would like to keep this deterministic error message.

theotherjimmy reviewed (changes requested):
I think this approach, allowing regions within a build to overlap, will prevent us from checking conditions that should fail. Instead, we could create a way to have multiple regions refer to the same file, and keep these checks

naveenkaje commented:
Build log from various build configurations attached
1. app at 0x16000, mbed-bootloader (at 0x64000)
2. bootloader and app are contiguous
3. Trying to place application within bootloader space. Should fail
[BuildLogs.txt](https://github.com/ARMmbed/mbed-os/files/2770890/BuildLogs.txt)

adbridge commented:
@theotherjimmy Can you please review the latest changes

adbridge commented:
@bridadan @JanneKiiskila would you guys also like to review ?

JanneKiiskila commented:
@LiyouZhou might be worth having a look?

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -792,26 +794,40 @@ def _generate_bootloader_build(self, rom_memories):
+                    else:
+                        break
+                    if end_address == None:
+                        raise ConfigException("bootloader segments don't fit within rom region")
```
Please remove "region" from this exception, as ROM is not a region:
```suggestion
                        raise ConfigException("bootloader segments don't fit within rom")
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -792,26 +794,40 @@ def _generate_bootloader_build(self, rom_memories):
+            else:
+                _, end_address = part.segments()
+                if (end_address > rom_end):
+                    raise ConfigException("bootloader segments don't fit within rom region")
```
Please remove "region" from this exception, as ROM is not a region:
```suggestion
                    raise ConfigException("bootloader segments don't fit within rom")
```

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -792,26 +794,40 @@ def _generate_bootloader_build(self, rom_memories):
+                    raise ConfigException("bootloader segments don't fit within rom region")
+                part_size = Config._align_ceiling(end_address, self.sectors) - rom_start
+                yield Region("bootloader", rom_start, part_size, False,
                          filename)
```
Could you re-indent this line?

naveenkaje replied:
bootloader regions are generated (bootloader0, bootloader1, so on) within the loop. I hope that's clear. Added a comment.

theotherjimmy replied:
The indentation is off. I'm just asking you to line up `filename` with the fist `"` in the previous line.

theotherjimmy commented on tools/config/__init__.py:
```diff
@@ -792,26 +794,40 @@ def _generate_bootloader_build(self, rom_memories):
+                        raise ConfigException("bootloader segments don't fit within rom region")
+                    part_size = Config._align_ceiling(end_address, self.sectors) - rom_start
+                    yield Region("bootloader"+`part_count`, start, part_size, False,
+                             filename)
```
This indentation looks off. Could you re-indent?

naveenkaje replied:
updated indentation in the new patch-set.

theotherjimmy reviewed (changes requested):
Style and UI nits.

theotherjimmy commented:
@naveenkaje Do you think we could have a few tests of the behavior we're expecting here?

LiyouZhou commented:
can the header be moved freely as well?

LiyouZhou commented:
> Tested with cloud client.
> 
> Log from build where BOOTLOADER for NRF52 is placed at 0x64000
> 
> ```
> 0 2408
> 409600 478028
> New offset for placing application 4096 last_address 409600 size 405504
> rom_end 524288 rom_end - start 520192
> Using ROM regions bootloader, header, application in this build.
>   Region bootloader: size 0x75000, offset 0x0
>   Region header: size 0x70, offset 0x75000
>   Region application: size 0x63000, offset 0x1000
> ```
why does it say here region bootloader is at offset 0x0 and if what is printed here is correct, it looks like the application overlapped with bootloader and the bootloader is not at the offset you say it is.

theotherjimmy commented:
>  why does it say here region bootloader is at offset 0x0 

That's pretty straight forward: because you have part of your bootloader at 0. You may call it the MBR, but the tools don't know (or care) about that naming convention.

> and if what is printed here is correct, it looks like the application overlapped with bootloader

@naveenkaje I was under the impression that this would generate more than one bootloader region. If this is true then @LiyouZhou are you using the code from the most recent version of this branch?

> and the bootloader is not at the offset you say it is

This is addressed above.

LiyouZhou commented:
Oh I was just reading this output from the top of this PR.

theotherjimmy commented:
@naveenkaje Could you show us what the new version looks like?

Commit:
tools: support cases where bootloader is in chunks

Support the requirement where bootloader can be in chunks and enable
placing the application at a particular offset specified by config.
With FEATURE_BOOTLOADER support, the bootloader can be placed at a
high address. Add support to the tools so that application can be
placed in the available space before the beginning of the bootloader.

Commit:
tools: build_api_test: Add tests to verify the processing of bootloader images

Add tests to
    1. Verify that a ConfigException is generated if application is placed
        within the bootloader region
    2. Verify that a ConfigException is generated if bootloader segments
        don't fit witin rom.

naveenkaje commented:
Build logs from
 1. mbed-bootlader and
 2. client-lite built with the bootloader binary built in step 1
```
Bootloader at 0x60000 with MBR
Scan: mbed-bootloader
Scan: env
Using ROM regions bootloader, application in this build.
  Region bootloader: size 0x1000, offset 0x0
  Region application: size 0x20000, offset 0x60000
....
```
mbed-client placed at 0x30000. Overlap is expected.
```
Building project client-lite (NRF52_DK, GCC_ARM)
Scan: client-lite
Using ROM regions bootloader1, bootloader2, header, application in this build.
  Region bootloader1: size 0x1000, offset 0x0
  Region bootloader2: size 0x68000, offset 0x60000
  Region header: size 0x70, offset 0x68000
  Region application: size 0x50000, offset 0x30000
Compile [  0.1%]: m2mtimer.cpp
...
Link: client-lite_application
Elf2Bin: client-lite_application
Merging Regions
  Filling region bootloader1 with /home/navkaj01/code/client-lite/configs/../tools/mbed-bootloader-nrf52_dk.hex
  Filling region header with ./BUILD/NRF52_DK/GCC_ARM/client-lite_header.hex
  Filling region application with ./BUILD/NRF52_DK/GCC_ARM/client-lite_application.hex
[ERROR] Data overlapped at address 0x60000
[mbed] ERROR: "/usr/bin/python" returned error.
       Code: 1
       Path: "/home/navkaj01/code/client-lite"
       Command: "/usr/bin/python -u /home/navkaj01/code/client-lite/mbed-os/tools/make.py -t GCC_ARM -m NRF52_DK --source . --build ./BUILD/NRF52_DK/GCC_ARM -c --app-config configs/wifi_esp8266_v4.json"
       Tip: You could retry the last command with "-v" flag for verbose output
```
mbed-client placed now at 0x10000, no overlap.
```
Building project client-lite (NRF52_DK, GCC_ARM)
Scan: client-lite
Using ROM regions bootloader1, bootloader2, header, application in this build.
  Region bootloader1: size 0x1000, offset 0x0
  Region bootloader2: size 0x68000, offset 0x60000
  Region header: size 0x70, offset 0x68000
  Region application: size 0x70000, offset 0x10000
Compile [  0.1%]: m2mtimer.cpp
Compile [  0.2%]: protoman_error_parser.c
<snip>
Link: client-lite_application
Elf2Bin: client-lite_application
Merging Regions
  Filling region bootloader1 with /home/navkaj01/code/client-lite/configs/../tools/mbed-bootloader-nrf52_dk.hex
  Filling region header with ./BUILD/NRF52_DK/GCC_ARM/client-lite_header.hex
  Filling region application with ./BUILD/NRF52_DK/GCC_ARM/client-lite_application.hex
Space used after regions merged: 0x68070
Merging Regions
  Filling region application with ./BUILD/NRF52_DK/GCC_ARM/client-lite_application.hex
Space used after regions merged: 0x3a590
| Module                              |           .text |       .data |          .bss |
|-------------------------------------|-----------------|-------------|---------------|
| [fill]                              |       438(+438) |     24(+24) |     127(+127) |
| [lib]/c.a                           |   50523(+50523) | 2474(+2474) |       89(+89) |
| [lib]/gcc.a                         |     4148(+4148) |       0(+0) |         0(+0) |
| [lib]/m.a                           |           8(+8) |       0(+0) |         0(+0) |
| [lib]/misc                          |       208(+208) |     12(+12) |       28(+28) |
| [lib]/nosys.a                       |         32(+32) |       0(+0) |         0(+0) |
| [lib]/stdc++.a                      |           1(+1) |       0(+0) |         0(+0) |
| main.o                              |     3089(+3089) |       4(+4) |       36(+36) |
| mbed-cloud-client/mbed-client       |   43558(+43558) |     29(+29) |       54(+54) |
| mbed-cloud-client/source            |     5154(+5154) |       9(+9) |       20(+20) |
| mbed-cloud-client/update-client-hub |   18005(+18005) |   232(+232) |   5150(+5150) |
| mbed-os/cmsis                       |     1033(+1033) |       0(+0) |       84(+84) |
| mbed-os/components                  |   12429(+12429) |       0(+0) |   2928(+2928) |
| mbed-os/drivers                     |     4489(+4489) |       0(+0) |       48(+48) |
| mbed-os/events                      |     1897(+1897) |       0(+0) |   2180(+2180) |
| mbed-os/features                    |   54861(+54861) |   157(+157) |     937(+937) |
| mbed-os/hal                         |     1859(+1859) |       9(+9) |     128(+128) |
| mbed-os/platform                    |     5331(+5331) |   276(+276) |   2705(+2705) |
| mbed-os/rtos                        |   12460(+12460) |   168(+168) |   6041(+6041) |
| mbed-os/targets                     |   14160(+14160) |     49(+49) |     726(+726) |
| mbed_cloud_dev_credentials.o        |       309(+309) |       0(+0) |         0(+0) |
| source/application_init.o           |       104(+104) |       0(+0) |         0(+0) |
| source/blinky.o                     |     1107(+1107) |       1(+1) |         4(+4) |
| source/oma_lwm2m_object_defs.o      |       338(+338) |       0(+0) |         0(+0) |
| source/platform                     |     1363(+1363) |       4(+4) |     103(+103) |
| source/resource.o                   |       300(+300) |       0(+0) |         0(+0) |
| update_default_resources.o          |         41(+41) |       0(+0) |         0(+0) |
| update_ui_example.o                 |       282(+282) |       0(+0) |         4(+4) |
| Subtotals                           | 237527(+237527) | 3448(+3448) | 21392(+21392) |
Total Static RAM memory (data + bss): 24840(+24840) bytes
Total Flash memory (text + data): 240975(+240975) bytes

```

naveenkaje commented:
> @naveenkaje Could you show us what the new version looks like?

@theotherjimmy @LiyouZhou 
I posted build logs from various scenarios. Let me know if you want me to try any other combinations.

theotherjimmy commented:
@naveenkaje It looks like those are now out of order. Could we sort them by start address before printing? Also, it looks like we don't print "bootloader2" ever when merging at the end. Could we print skipped as it's already present or something?

naveenkaje commented:
Updated with new patch-set addressing
1. indentation
2. sorting regions as per addresses before printing/logging
3. printing/logging skipped regions when merged previously.

naveenkaje commented:
Fixed error raised by CI

```
tools/toolchains/__init__.py:43: in <module>
    from ..config import (ConfigException, RAM_ALL_MEMORIES, ROM_ALL_MEMORIES)
E     File "/home/travis/build/ARMmbed/mbed-os/tools/config/__init__.py", line 823
E       yield Region("bootloader"+`part_count`, start, part_size, False, filename)
E                                 ^
E   SyntaxError: invalid syntax
```
with
```
yield Region("bootloader"+str(part_count), start, part_size, False, filename)
```

naveenkaje commented:
Fixed the following error flagged by CI

```
        """
            Test that merge region fails as expected when part size overflows region size.
            """
>       tgt = unicode('\x80NRF52_DK', errors='ignore')
E       NameError: name 'unicode' is not defined
```

cmonr commented:
@naveenkaje Looks like there's still one more Py3 issue that needs to be fixed.

bridadan reviewed (changes requested):
The code is looking good so far! And thanks very much for the test cases.

Your last log output is confusing since the offset+sizes seem to overlap. Perhaps the sizes aren't being computed correctly anymore?

bridadan commented on tools/config/__init__.py:
```diff
@@ -780,13 +780,29 @@ def _make_header_region(self, start, header_format, offset=None):
+        end_address = rom_end
+
+        for s, e in region_list:
+            if start_address < s and start_address < e:
```
Wouldn't the `start_address < e` be unnecessary? The "end" address (`e`) should always be greater than the "start" address (`s)` right?

bridadan commented on tools/config/__init__.py:
```diff
@@ -780,13 +780,29 @@ def _make_header_region(self, start, header_format, offset=None):
+        """Given a start address and set of regions, compute
+        the end address. The end address is either
+        rom_end or beginning of next section, whichever
+        is smaller
```
We should document that we expect the `region_list` to be sorted based on start address. OR you can always sort the regions in this function to ensure that.

naveenkaje replied:
Good point. Uploaded a patch set with suggested changes. Please re-review

bridadan commented on tools/config/__init__.py:
```diff
@@ -780,13 +780,29 @@ def _make_header_region(self, start, header_format, offset=None):
+        for s, e in region_list:
+            if start_address < s and start_address < e:
+                end_address = s
+        return end_address
```
This will continue looping even once the valid region is found, I think you want something closer to this:

```python
for s, e in region_list
    if start_address < s:
        return s
return end_address
```

bridadan reviewed (dismissed):
Love that new log output! So slick! 😄 

I have a few concerns about the `_get_end_address` function.
